mlxbf-bootctl.c

```diff
@@ -753,16 +755,19 @@ bool should_install_image(boot_image_header_t *hdr, int version, uint8_t *max_ve
   if (max_versions == NULL)
     die("internal: max_versions is NULL");
 
-  bool should_not_install = (
-    // Do not install anything introduced after given version
-    hdr->cur_img_ver > version ||
+  // The logic to decide what to install and what not to install
+  // is based on BlueField 1, 2, 3, <X> chips each possibly having
+  // their own version of a specific image while at the same time
+  // also using "shared" images.  A shared image is defined as an
+  // image with a version of 0 AND there are no other higher
+  // versions in the bootstream (i.e. images with a version > 0).
 
-    // Otherwise, check for redundancy.
-    // If the img version isn't the max, it shouldn't be installed.
-    hdr->cur_img_ver < max_versions[hdr->image_id]
-  );
+  // If the version matches OR the image is a shared image, install it.
+  if ((hdr->cur_img_ver == version) || (max_versions[hdr->image_id] == 0)) {
+    install = true;
```

Comment:
I understand the problem of the original commit now.  Line 802 below missed a condition check that it should be
"if (hdr.cur_img_ver <= version && hdr.cur_img_ver > max_versions[hdr.image_id])", so max_versions[] doesn't  include images with higher version. It's the logic of UEFI capsule which works I think.

UEFI doesn't such such issue (line 911-912) which should work well I think.

 906   // First loop to decide the best version
 907   BytesLeft = BufferSize;
 908   Index = (UINT8*) Buffer;
 909   while (BytesLeft > 0) {
 910     Header = (BOOT_IMAGE_HEADER*) Index;
 911     if (Header->data.cur_img_ver <= ImageVer &&
 912         Header->data.cur_img_ver > ImageMaxVer[Header->data.image_id]) {
 913         ImageMaxVer[Header->data.image_id] = Header->data.cur_img_ver;
 914     }
 915     ImageSize = PmiBootHeaderGetImageSize (Header);
 916     Index += ImageSize;
 917     BytesLeft -= ImageSize;
 918   }

However the new logic also seem good.

Reply:
I will need to review the capsule code to understand better.